lib/Analysis/Utility.cpp

```diff
@@ -149,9 +149,29 @@ bool supportMMA(triton::DotOp op, int version) {
 }
 
 #ifdef USE_ROCM
+static bool supportMFMAGranularity(int dim_size) {
+  std::vector<int> supported_granularity{32};
+  for (int granularity: supported_granularity)
+    if (dim_size % granularity == 0)
+      return true;
+  return false;
+}
+
 bool supportMFMA(triton::DotOp op) {
-  auto aElemTy = op.getA().getType().cast<RankedTensorType>().getElementType();
-  auto bElemTy = op.getB().getType().cast<RankedTensorType>().getElementType();
+  auto aTy = op.getA().getType().cast<RankedTensorType>();
+  auto bTy = op.getB().getType().cast<RankedTensorType>();
+
+  auto aShape = aTy.getShape();
+  auto bShape = bTy.getShape();
+
+  assert(aShape[1] == bShape[0]);
+  if (!supportMFMAGranularity(aShape[0]) ||
+      !supportMFMAGranularity(aShape[1]) ||
+      !supportMFMAGranularity(bShape[1]))
+    return false;
+
+  auto aElemTy = aTy.getElementType();
+  auto bElemTy = bTy.getElementType();
```

Comment:
This part enables fallback to FMA implementation for small matrix sizes in C++ part of compiler